tests/test_stencil_arithmetic.py

```diff
@@ -2,66 +2,129 @@
-def x(xdim=4):
-    return Dimension(name='x', size=xdim)
+def i(size=3):
+    return Dimension(name='i', size=size)
 
 
 @pytest.fixture
-def y(ydim=6):
-    return Dimension(name='y', size=ydim)
+def j(size=4):
+    return Dimension(name='j', size=size)
 
 
 @pytest.fixture
-def ai(x, y, name='a', value=2.):
-    a = DenseData(name=name, dimensions=(x, y))
-    a.data[:] = value
-    return a.indexify()
+def k(size=5):
+    return Dimension(name='k', size=size)
 
 
 @pytest.fixture
-def bi(x, y, name='b', value=3.):
-    b = DenseData(name=name, dimensions=(x, y))
-    b.data[:] = value
-    return b.indexify()
+def l(size=6):
+    return Dimension(name='l', size=size)
+
+
+def symbol(name, dimensions, value=0., mode='function'):
+    """Short-cut for symbol creation to test "function"
+    and "indexed" API."""
+    assert(mode in ['function', 'indexed'])
+    s = DenseData(name=name, dimensions=dimensions)
+    s.data[:] = value
+    return s.indexify() if mode == 'indexed' else s
 
 
 @pytest.mark.parametrize('expr, result', [
-    ('Eq(ai, ai + bi + 5.)', 10.),
-    ('Eq(ai, bi - ai)', 1.),
-    ('Eq(ai, 4 * (bi * ai))', 24.),
-    ('Eq(ai, (6. / bi) + (8. * ai))', 18.),
+    ('Eq(a, a + b + 5.)', 10.),
+    ('Eq(a, b - a)', 1.),
+    ('Eq(a, 4 * (b * a))', 24.),
+    ('Eq(a, (6. / b) + (8. * a))', 18.),
 ])
-def test_arithmetic_flat(ai, bi, expr, result):
+@pytest.mark.parametrize('mode', ['function', 'indexed'])
+def test_arithmetic_flat(i, j, expr, result, mode):
     """Tests basic point-wise arithmetic on two-dimensional data"""
+    clear_cache()
+    a = symbol(name='a', dimensions=(i, j), value=2., mode=mode)
+    b = symbol(name='b', dimensions=(i, j), value=3., mode=mode)
+    fa = a.base.function if mode == 'indexed' else a
+    fb = b.base.function if mode == 'indexed' else b
+
     eqn = eval(expr)
-    StencilKernel(eqn)(ai.base.function, bi.base.function)
-    assert np.allclose(ai.base.function.data, result, rtol=1e-12)
+    StencilKernel(eqn)(fa, fb)
+    assert np.allclose(fa.data, result, rtol=1e-12)
 
 
 @pytest.mark.parametrize('expr, result', [
-    ('Eq(ai, ai + bi + 5.)', 10.),
-    ('Eq(ai, bi - ai)', 1.),
-    ('Eq(ai, 4 * (bi * ai))', 24.),
-    ('Eq(ai, (6. / bi) + (8. * ai))', 18.),
+    ('Eq(a, a + b + 5.)', 10.),
+    ('Eq(a, b - a)', 1.),
+    ('Eq(a, 4 * (b * a))', 24.),
+    ('Eq(a, (6. / b) + (8. * a))', 18.),
 ])
-def test_arithmetic_deep(expr, result):
+@pytest.mark.parametrize('mode', ['function', 'indexed'])
+def test_arithmetic_deep(i, j, k, l, expr, result, mode):
     """Tests basic point-wise arithmetic on multi-dimensional data"""
-    i = Dimension(name='i', size=3)
-    j = Dimension(name='j', size=4)
-    k = Dimension(name='k', size=5)
-    l = Dimension(name='l', size=6)
-
-    a = DenseData(name='a', dimensions=(i, j, k, l))
-    a.data[:] = 2.
-    ai = a.indexify()
-    b = DenseData(name='b', dimensions=(j, k))
-    b.data[:] = 3.
-    bi = b.indexify()
+    clear_cache()
```

Comment:
Ok, this originally fixed a suspected caching bug, but after several re-writes it now seems fine without the fix. Good catch.